### PR TITLE
Genesis

### PR DESCRIPTION
**Pull request by mattruggio**
This PR includes the initial implementation of three data structures:

* HashMath::Matrix
* HashMath::Record
* HashMath::Table

See the README for examples for each, but each is essentially a higher-level Hash data structure.

### CODE REVIEW

File: lib/hash_math/record.rb [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+      @prototype = keys.map { |key| [key, base_value] }.to_h
+
+      freeze
+    end
+
+    def make!(hash = {})
+      make(hash, true)
+    end
+
+    def make(hash = {}, bound = false)
```
[the remark this answers is not in the document]
Comment (mattruggio) [reply lib/hash_math/record.rb]: Fixed

[the remark this answers is not in the document]
Comment (mattruggio) [reply lib/hash_math/record.rb]: Fixed

File: README.md [hunk 2]
```diff
@@ -2,4 +2,238 @@
 
 [![Gem Version](https://badge.fury.io/rb/hash_math.svg)](https://badge.fury.io/rb/hash_math) [![Build Status](https://travis-ci.org/bluemarblepayroll/hash_math.svg?branch=master)](https://travis-ci.org/bluemarblepayroll/hash_math) [![Maintainability](https://api.codeclimate.com/v1/badges/9f9a504b3f5df199a253/maintainability)](https://codeclimate.com/github/bluemarblepayroll/hash_math/maintainability) [![Test Coverage](https://api.codeclimate.com/v1/badges/9f9a504b3f5df199a253/test_coverage)](https://codeclimate.com/github/bluemarblepayroll/hash_math/test_coverage) [![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
 
-Under Construction.
+Ruby's Hash data structure is ubiquitous and highly flexible.  This library contains common Hash patterns widely used within our code-base:
+
+* Matrix: build a hash up then expand into hash products
+* Record: define a list of keys and base value to use as a blueprint for hashes
+* Table: key-value pair two-dimensional array of hash builder
```
Comment (gerryster) [inline README.md]: How about ending this with "hash _builders_"?

File: README.md [hunk 3]
```diff
@@ -2,4 +2,238 @@
...
+]
+````
+
+Notes:
+
+* keys are type-sensitive and works just like Hash keys work.
+
+### Table: The Double Hash (Hash of Hashes)
+
+HashMath::Table builds on top of HashMath::Record.  It constructs a table data-structure using a key-value pair builder method: `#add(row_id, field_id, key, value)`.  It ultimately outputs an array of Row objects where each Row object has a row_id and fields (field_id => value) hash.  The value proposition here is you can iterate over a key-value pair and construct each row any way you wish.
```
Comment (gerryster) [inline README.md]: I am a little confused here as this implies that there are four arguments to #add but the example below only passes in three and the third one looks like the value.

File: README.md [hunk 4]
```diff
@@ -2,4 +2,238 @@
 
 [![Gem Version](https://badge.fury.io/rb/hash_math.svg)](https://badge.fury.io/rb/hash_math) [![Build Status](https://travis-ci.org/bluemarblepayroll/hash_math.svg?branch=master)](https://travis-ci.org/bluemarblepayroll/hash_math) [![Maintainability](https://api.codeclimate.com/v1/badges/9f9a504b3f5df199a253/maintainability)](https://codeclimate.com/github/bluemarblepayroll/hash_math/maintainability) [![Test Coverage](https://api.codeclimate.com/v1/badges/9f9a504b3f5df199a253/test_coverage)](https://codeclimate.com/github/bluemarblepayroll/hash_math/test_coverage) [![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
 
-Under Construction.
+Ruby's Hash data structure is ubiquitous and highly flexible.  This library contains common Hash patterns widely used within our code-base:
+
+* Matrix: build a hash up then expand into hash products
+* Record: define a list of keys and base value to use as a blueprint for hashes
```
[the remark this answers is not in the document]
Comment (mattruggio) [reply README.md]: It should be "keys and a base value"

File: README.md (hunk 3 above)
Comment (mattruggio) [reply README.md]: row_id is the Y-axis, field_id is the X-axis; key should be removed.

File: README.md (hunk 2 above)
Comment (mattruggio) [reply README.md]: I changed this, see if the new wording is better.

**Comment by gerryster** [comment]
LGTM!

**Pull request merged**